src/VisualStudio/Core/Def/LanguageService/PackageRegistrationTasks.cs

```diff
@@ -0,0 +1,90 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.VisualStudio.Shell;
+using Microsoft.VisualStudio.Threading;
+
+namespace Microsoft.VisualStudio.LanguageServices.Implementation.LanguageService;
+
+using WorkTask = Func<IProgress<ServiceProgressData>, PackageRegistrationTasks, CancellationToken, Task>;
+
+/// <summary>
+/// Provides a mechanism for registering work to be done during package initialization. Work is registered
+/// as either main thread or background thread appropriate. This allows processing of these work items
+/// in a batched manner, reducing the number of thread switches required during the performance sensitive
+/// package loading timeframe.
+/// 
+/// Note that currently the processing of these tasks isn't done concurrently. A future optimization may
+/// allow parallel background thread task execution, or even concurrent main and background thread work.
+/// </summary>
+internal sealed class PackageRegistrationTasks(JoinableTaskFactory jtf)
+{
+    private readonly List<WorkTask> _backgroundThreadWorkTasks = [];
+    private readonly List<WorkTask> _mainThreadWorkTasks = [];
+    private readonly JoinableTaskFactory _jtf = jtf;
+    private readonly object _gate = new();
+
+    public void AddTask(bool isMainThreadTask, WorkTask task)
+    {
+        // This lock is a bit extraneous, as the current code doesn't register or process tasks concurrently.
+        lock (_gate)
+        {
+            var workTasks = GetWorkTasks(isMainThreadTask);
+            workTasks.Add(task);
+        }
+    }
+
+    public async Task ProcessTasksAsync(IProgress<ServiceProgressData> progress, CancellationToken cancellationToken)
+    {
+        // prime the pump by doing the first group of bg thread work if the initiating thread is not the main thread
+        if (!_jtf.Context.IsOnMainThread)
+            await PerformWorkAsync(isMainThreadTask: false, progress, cancellationToken).ConfigureAwait(false);
+
+        // Continue processing work until everything is completed, switching between main and bg threads as needed.
+        while (_mainThreadWorkTasks.Count > 0 || _backgroundThreadWorkTasks.Count > 0)
```

Comment:
This is technically using the lists outside a lock; it's _probably_ fine unless some CPU cache shenanigans happen (i.e. the CPU core it's running on never sees the write).

Reply:
switched to just use a concurrent data structure instead of trying to handle the concurrent access right using Lists